pkgs/dart_model/lib/src/json_buffer.dart

```diff
@@ -0,0 +1,390 @@
+import 'dart:collection';
+import 'dart:convert';
+import 'dart:typed_data';
+
+/// Map with lazily evaluated values.
+///
+/// This is the efficient way to accumulate data into a [JsonBuffer]. Nested
+/// maps should also be `LazyMap`.
+class LazyMap with MapMixin<String, Object?> implements Map<String, Object?> {
+  @override
+  final List<String> keys;
+  final Object? Function(String) lookup;
+
+  LazyMap(Iterable<String> keys, this.lookup)
+      : keys = UnmodifiableListView(keys.toList());
+
+  @override
+  Object? operator [](Object? key) => key is String ? lookup(key) : null;
+
+  @override
+  void operator []=(String key, Object? value) =>
+      throw UnsupportedError('LazyMap is immutable.');
+
+  @override
+  void clear() => throw UnsupportedError('LazyMap is immutable.');
+
+  @override
+  Object? remove(Object? key) =>
+      throw UnsupportedError('LazyMap is immutable.');
+}
+
+/// Bytebuffer-backed JSON data.
+///
+/// Data can be accumulated directly into the buffer which is then immediately
+/// ready to write to the wire. On the receiving side, the data can be
+/// accessed directly as a `Map<String, Object?>` without copying out of the
+/// buffer.
+///
+/// The constructor takes a [Map], but passing in a normal `Map` _does_
+/// require copying. Instead, instantiate with a [LazyMap]. The lazy map will
+/// then be evaluated into the buffer. Nested maps in the `LazyMap` must _also_
+/// use `LazyMap` otherwise it will still be necessary to copy those.
+//
+// TODO(davidmorgan): check if it's worth adding a `LazyList` for the same
+// reason.
+// TODO(davidmorgan): this is just a proof of concept to make sure `dart_model`
+// is compatible with fast JSON. Write up a design for discussion, complete
+// the implementation.
+class JsonBuffer {
+  // TODO(davidmorgan): _seenStrings and _decodedStrings are some simple
+  // optimizations; complete the design and implementation.
+  final Map<String, _Pointer> _seenStrings = {};
+  final Map<_Pointer, String> _decodedStrings = {};
+
+  /// The JSON data.
+  // TODO(davidmorgan): what's a good initial size?
+  Uint8List _buffer = Uint8List(32);
```

Comment:
Do we know how many entries there will be in each map? I think we probably could.

If so, and also everything is either a pointer or an int/bool, then we could know exactly what size to always allocate this to, I think?

I am not totally certain though, and of course optional values make that more interesting.

Reply:
Or, just use a `BytesBuilder` instead of manually growing this ourselves? I think it basically does all the logic you are doing here.

Reply:
The maps are stored as key and value pointers, and we know how many entries there will be if it's a `LazyMap`, but the keys and values also go into the buffer, along with all the nested maps, and we don't know how big those will be.

`BytesBuilder` indeed looks similar; but it doesn't allow accesses as you're building, while this does. Not sure if that matters, I'll add a TODO to consider `BytesBuilder`.